ingestion_server/ingestion_server/cleanup.py

```diff
@@ -93,7 +113,82 @@ def _cleanup_tags(tags):
 }
 
 
-def _clean_data_worker(rows, temp_table, providers_config):
+class TlsTest:
+    """
+    URLs crawled from upstream are often lacking protocol information, or
+    use HTTP when HTTPS is available. We have to test a small sample of the
+    URLs to determine what protocol should be appended to each URL in the
+    event that it is missing or incorrect.
+    """
+    @classmethod
+    def _test_tls_supported(cls, url):
+        # No protocol provided
+        if 'https://' not in url and 'http://' not in url:
+            fixed_url = 'http://' + url
+            return cls._test_tls_supported(fixed_url)
+        # HTTP provided, but we want to check if HTTPS is supported as well.
+        elif 'http://' in url:
+            https = url.replace('http://', 'https://')
+            try:
+                res = re.get(https, timeout=2)
+                log.info('{}:{}'.format(https, res.status_code))
+                return 200 <= res.status_code < 400
+            except re.RequestException:
+                return False
+        # If HTTPS is in the URL already, we're going to trust that HTTPS is
+        # supported.
+        else:
+            return True
+
+    @staticmethod
+    def test_provider_tls_images_available(table):
+        """
+        Given a table, find 10 sample images and test whether HTTPS is
+        available. If the majority supports TLS, we assume TLS is supported by
+        the provider for all items.
+
+        :return: A dict with key "provider" mapped to value True if TLS is
+        available, else False.
+        """
+        provider_tls_supported = {}
+        conn = database_connect()
+        cur = conn.cursor(cursor_factory=DictCursor)
+        provider_query = 'SELECT provider_identifier FROM content_provider;'
+        cur.execute(provider_query)
+        providers = cur.fetchall()
+
+        for p in providers:
+            p = p[0]
+            sample_query = "SELECT thumbnail, url FROM temp_import_{table}" \
+                           " WHERE provider='{provider}' LIMIT 10" \
```

Comment:
would it make sense to pick 10 random items to have a better chance of finding anomalies?

Reply:
We could pick 10 random items, but it would be extremely slow since the temporary table doesn't have any indexes (by design, because bulk updating an indexed table is really slow). It also may result in the TLS supported status changing randomly when we ingest if a provider has a mix of HTTP and HTTPS links. I'd like to leave it at 10 for now, especially since this code will be disabled once we start generating thumbnails ourselves.

Reply:
sounds good